### ISSUE TITLE
Files and Folder conflicts

### ISSUE THREAD

**Issue by dk-zero-cool** [rclone/rclone#6264]
Hi. 

I am using rclone to keep a local server in sync with a remote server, but I ran into an issue a few days ago. I made some changes to a directory, where I split a file that had grown in size, into multiple files. These files were added to a directory with the same name as the original file that they derived from. This caused a conflict in my rclone sync, as it refused to replace the remote file with my new local directory. I took this into the [forum for help](https://forum.rclone.org/t/sync-not-working-properly/31318). 

Here are the highlights from that: 
 > Rclone can't and won't overwrite a directory with a file - that could lead to data loss.

__My take on this is:__ 

On purpose or not, I do still see this as a bug.

What I am trying to run is not copy but sync. As the docs itself states:

```
Sync (one way) mode to make a directory identical
```

A `one way` mode to make a directory `identical`. In such a mode everything in the target location should be completely discarded, unless it already matches the source. When using the sync command, you implicitly tell rclone that you do not care about the target. All that you want, is to have it match the source.

And there is no safety in this behavior. Many files does not per definition mean important files. A single file can easily be magnitudes more important than a directory with multiple files. Yet is has no problem deleting a single file that does not exist in the source.

**Comment by ncw** [comment]
I'm sure there is an existing issue about this with more discussion - I can't find it right now.

I'll just note that this commit made the situation worse by allowing a file and a directory with the same name, which pushed the checking into the backend which in turn leads to much worse error condtions.

https://github.com/rclone/rclone/commit/4b27c6719b636dbafde4fea03e47915ec7c78a3c

If we want to fix this issue then we'll have to revert that commit or add some infrastructure to tell whether a backend can have a file and a directory with the same name.

**Comment by dk-zero-cool** [comment]
What about just adding an option to the sync mode and have that be disabled by default? Then people can adapt manually to the type of implementation that they have? 

Also I want to point out again, that this is a sync option. In no scenario would a sync harm any actual data, as nothing in a one-way sync would ever touch the source. The idea is, like the manual states, to make a remote location identical (a mirror) to a source. If I change a file into a folder on the source location, then it only makes sense to make those same changes on the remote. Otherwise what is the real difference between the `copy` and `sync` options?

**Comment by ncw** [comment]
Here are some of the complications which need to be considered. None of these are insurmountable however they need careful consideration.

I think disobeying the user if they've set `--delete-after` (the default) might be the biggest sticking point. We don't usually delete data until the sync is complete.

## overwriting a file with a directory

- we need to delete the file **first** as we can't overwrite a file with a directory on most backends
    - this means disobeying the users `--delete-after`
- this is easy to back up with `--backup-dir` / `--suffix` at least

## overwriting a directory with a file

- we need to delete the directory **first** as we can't overwrite a directory with a file on most backends
    - this means disobeying the users `--delete-after`
- deleting a directory is not the same as deleting a file on rclone backends.
    -  We have the `Purge` method which does this which has native implementations on some backends.
    - `operations.Purge` implements the delete a directory recursively with fallback to native.
- With `--backup-dir` / `--suffix` deleting a directory means we need to move the directory to the backup
    - some backends support MoveDir which will make this easy
    - the rest don't which will require recursing to use Move/Copy on each file
    - `operations.DirMove` implements the two items above already so we can use that

**Comment by dk-zero-cool** [comment]
I don't agree with the fact that it would disobey `--delete-after`, depending of cause how you look at this option. 

1. If you look at this as a simple object with a name. When that object is a file, you simply replace one file with another _(modified content only, but that is mostly due to reducing the overhead)_. So one object replaces another object with the same name, and thereby the `--delete-after` is ignored as it does not apply. Why should this be any different when one object is a directory? It's still the same. You have two different objects with the same name. You want to make one identical to another. Thereby `--delete-after` does not apply, as no object has been deleted. One object has simply been altered compared to an existing one. 
2. There are 3 different delete options. There is no reason why these should act the same. If so, why have 3? If the `--delete-after` creates a conflict, simply report this in the log, skip and move on. The user can then decide if this is intentional or make use of one of the other 2 delete options to resolve it.

**Comment by ncw** [comment]
Your option 2 sounds like a good one to me.



> But then again, it seams that some implementations allows for files and folders to share a common name. That does of cause complicate things a bit. Not familiar with such implementations, but they sound illogical.

Amazon s3/google cloud storage/azure blob/b2 all allow a file and a directory with the same name.

Google Drive allows many files to have the same name (cf `rclone dedupe`).

In general rclone requires its backends to store files and directories more or less like a local file system, so I think merging 4b27c67 was probably a mistake as it breaks that. Fine to have it as an option.

**Comment by privat-eco** [comment]
Hi we've run in similar use case where we wanted to `rclone sync` a bunch of git repositories from a git server. Here is the scenario : 

- a user creates a branch `features/12`  (so on the server we have `refs/heads/features/12` and `refs/heads/features/` is a folder)
- a `rclone sync` happens (so there is a folder `refs/heads/features/` on the remote
- a user, on the same repository, deletes its branch `features/12`, then creates a branch `features` (so on the server `refs/heads/features/` is now a folder folder)
- if we launch `rclone sync` again the following error happens:
  ```
  2026/02/01 02:00:06 ERROR : repositories/@hashed/20/49/20491a692875572f441df44f504753f51bfba766c258699fc1c04014dbb3020a.git/refs/heads/features.64930375.partial: partial file rename failed: can't move object 
  - incompatible remotes
  2026/02/01 02:12:03 ERROR : Local file system at /data/git-storage: not deleting directories as there were IO errors
  2026/02/01 02:12:03 ERROR : Attempt 2/3 failed with 1 errors and: can't move object - incompatible remotes
  2026/02/01 02:12:31 ERROR : repositories/@hashed/20/49/20491a692875572f441df44f504753f51bfba766c258699fc1c04014dbb3020a.git/refs/heads/features.64930375.partial: partial file rename failed: can't move object 
  - incompatible remotes
  ```

Our backends are sftp for the source, local for the destination, but we can reproduce this really easily with local folders too.

Here is a little lab to help to reproduce this:

```dockerfile
FROM alpine:latest
ARG RCLONE_VERSION=v1.73.0
RUN apk add wget bash

# Download a rclone release, put it under /rclone, symlink /rclone/rclone -> /usr/bin/rclone
RUN wget https://github.com/rclone/rclone/releases/download/$RCLONE_VERSION/rclone-$RCLONE_VERSION-linux-amd64.zip && \
    unzip -jd /rclone rclone-$RCLONE_VERSION-linux-amd64.zip && \
    ln -s /rclone/rclone /usr/bin/rclone

# Creates a basic structure which is actually impossible to sync
RUN  mkdir -p /test/dest/ /test/src/ && \
     touch /test/src/element && \
     mkdir /test/dest/element

# Prepare a welcoming message to the lovely one which will try to fix this :-)
RUN echo "**************************************" > /etc/motd && \
    echo "There is some test folders at /test/src and /test/dest." >> /etc/motd && \
    echo "inside these, /test/src/element is a file and /test/dest/element is a directory." >> /etc/motd && \
    echo "" >> /etc/motd && \
    echo "Try something like this and:" >> /etc/motd && \
    echo "rclone sync /test/src/ /test/dest/" >> /etc/motd && \
    echo "expected: rclone should produce an exact copy of /test/src/ on /test/dest (replacing the element folder by the element file)" && \
    echo "**************************************" >> /etc/motd
CMD ["login", "-f", "root"]
```

With this lab we want to sync `/test/src/` with `/test/dest/`. On src `element` is a file ; on dest `element` is a directory.

To use the lab:
1. save above content in a `Containerfile` or `Dockerfile`
1. `sudo docker build -t rclone-sync-debug .`
1. `sudo docker run --rm -it rclone-sync-debug`
1. `rclone sync /test/src/ /test/dest/`

You can try with another `rclone` release by adding at the docker build step the argument `--build-arg RCLONE_VERSION=vX.X.X` where `vX.X.X` is the release you want to give a try.

Another interesting thing:
- if `element` is a directory in src and a file in dest ; then `rclone sync` will `delete dest/element`.... and there will be NO element item in the dest directory **with no error/warning** (this means that in our case **a branch may be lost in the process of syncing**)
- if `element` is a file in src and a directory in dest ; then `rclone sync` will fail (as already stated in this issue)

hope this helps

**Comment by privat-eco** [comment]
Using the following args seems to make rclone sync work as expected : `--delete-before --create-empty-src-dirs`.

IMHO this should be the default behaviour
